### ISSUE TITLE
OWL-Vit : Questions regarding setting up data for fine-tuning

### ISSUE THREAD

**Issue by fizahkhalid** [google-research/scenic#655]
Hello, I have my own dataset which I would like to use for fine-tuning the detection model. 

I have text labels associated with bounding boxes in my images. I'm not sure how do I need to arrange my data to start with the fine-tuning pipeline. 

Also, I don't know what "crowd" , "non exhaustive labels" mean, and what do I put there. 

And is what does "NEGATIVE LABELS" and "NEGATIVE TEXT LABELS" have different. Is one of them mapping of the text label in integer form?

Thanks a lot for this awesome repo btw :')

**Comment by mjlm** [comment]
`CROWD` and `NOT_EXHAUSTIVE_LABELS` are defined as in [COCO](https://cocodataset.org/#format-data)/[LVIS](https://www.lvisdataset.org/dataset):

* `CROWD` is `True` for box annotations that contain more than one instance (e.g. a single box around a crowd of people). If your dataset does not have "crowd" annotations, you can use zero/False values.
* `NOT_EXHAUSTIVE_LABELS` is for federated datasets such as LVIS, which have the basic assumption that either all or no instances of a category are annotated in an image. `NOT_EXHAUSTIVE_LABELS` is a list of categories for which this assumption is violated, i.e. for which some instances in the image are not annotated (e.g. a large crowd of people where only some have a bounding box annotation). If your dataset doesn't have this, you can leave the list empty.

For the negative labels, you're right, `NEGATIVE_LABELS` contains integer labels and `NEGATIVE_TEXT_LABELS` contains the corresponding strings. For training, only `NEGATIVE_TEXT_LABELS` is required.

More explanations at https://github.com/google-research/scenic/blob/main/scenic/projects/owl_vit/preprocessing/modalities.py

**Comment by kopyl** [comment]
@fizahkhalid

hey :)
Were you able to fine-tune the OWL ViT model?)

**Comment by kopyl** [comment]
@mjlm do you know any places where I can find any OWL-ViT fine-tuning guides?

**Comment by fizahkhalid** [comment]
> @fizahkhalid
> 
> hey :) Were you able to fine-tune the OWL ViT model?)

I was, but I did not get good results since, maybe I was doing something wrong, or had to engineer the prompts a bit more, and my dataset is very different from the ones used to train the Clip model.

I couldn't find any guides when I was working on it, I'll try to look for them and let you know. 

Since I was interested in Vision Language models, and OWL ViT did not do well for me, I looked into this repository, https://github.com/zengyan-97/X-VLM, I have tried their object detection and Visual QA, the results were pretty amazing.

**Comment by kopyl** [comment]
@fizahkhalid thanks ❤️ 

I would like to know what was your use-case.
Can we talk in the DM? If you have Telegram, you can hit me @kopyl .

**Comment by loveshsama** [comment]
@fizahkhalid 

hey,
Since you were able to fine-tune, Can you help me the repo or finetune code, or step by step guide on how to finetune. I am unable to figure this out.

**Comment by fizahkhalid** [comment]
@loveshsama 

I can't go into much detail as I did this a while back and don't remember much.

First go to the file **configs/clip_b32_finetune.py**

Here you can adjust the parameters for your training, and you'll also come across the following:

```
DETECTION_FEATURES = ('boxes', 'crowd', 'image', 'instance_labels',
                      'instance_text_labels', 'negative_labels',
                      'negative_text_labels', '_seed', 'seed')
```
                      
 This gave me an idea that how my data should look like. I would advise you first, to try finetuning with any of the data the author's have already used for training, just to see if everything is working fine.
 
My images only had the bounding boxes and their labels, so after seeing the above "DETECTION FEATURES", I added the necessary fields in my dataset.

After setting up your data accordingly, and changing the parameters as per your requirement. 

Now, you can go to the data pre-processing code, and see if it aligns with your use-case.  
 
 I found the following files important:

**1.  label_ops
2. image_ops
3. input_pipeline**

I think, "ops" means "operations" and these files contain the augmentation operations etc for labels and images. And the last file controls the complete pipeline, the mosaic formation etc. 

Here I found out that we have to **build our dataset using TFDS.** You'll have to look into that. 

_Depending on the use-case, you'll have to maybe change some fields in the "label_ops" file, as it contains the "prompts templates" and the "rare cases" of the datasets._

**Comment by fizahkhalid** [comment]
@kopyl  

Left you a message on Telegram !

**Comment by WKaiH123** [comment]
@fizahkhalid Hello! My images are also just bounding boxes and their labels. According to "DETECTION FEATURES", please ask how you added the necessary fields in the dataset. Thanks!

**Comment by MattLiutt** [comment]
@fizahkhalid Hi, I'm also working on my own dataset, may I know how you configure your own dataset? Thanks!
